File: dairy_erp/dairy_erp/report/cattle_feed_sales_report/cattle_feed_sales_report.js
// Copyright (c) 2016, indictrans technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Cattle Feed Sales Report"] = {
	"filters": [
		{
			"fieldname":"vlcc",
			"label": __("VLCC"),
			"fieldtype": "Link",
			"options":"Company",
			"hidden": 1
		},
		{
			"fieldname":"farmer",
			"label": __("Farmer"),
			"fieldtype": "Link",
			"options":"Farmer",
			"get_query": function (query_report) {
				var vlcc_name = frappe.query_report_filters_by_name.vlcc.get_value();
				if(vlcc_name){
					return {
						"filters": {
							"vlcc_name": vlcc_name					
							}
						}
				}
			},
			"reqd":1,
			"on_change":function(query_report){
				frappe.call({
					method: "frappe.client.get_value",
					args: {
						doctype: "Farmer",
						filters: {"name": frappe.query_report_filters_by_name.farmer.get_value()},
						fieldname: ["farmer_id","full_name"]
					},
					callback:function(r){
						if(r.message){
							frappe.query_report_filters_by_name.farmer_id.set_input(r.message['farmer_id']);
							frappe.query_report_filters_by_name.full_name.set_input(r.message['full_name']);
							query_report.trigger_refresh();		
						}
						else{
							frappe.query_report_filters_by_name.farmer_id.set_input("sas");
							frappe.query_report_filters_by_name.full_name.set_input("asa");
							query_report.trigger_refresh();
						}		
					}
				})
			}
		},
		{
			"fieldname":"farmer_id",
			"label": __("Farmer id"),
			"fieldtype": "Data",
			"read_only":1
		},
		{
			"fieldname":"full_name",
			"label": __("Full Name"),
			"fieldtype": "Data",
			"read_only":1
		},
		{
			"fieldname":"start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			"reqd":1,
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname":"end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			"reqd":1,
			"default": frappe.datetime.get_today()
		}
	],
	onload: function(query_report) {
		frappe.call({
			method: "frappe.client.get_value",
			args: {
				doctype: "User",
				filters: {"name": frappe.session.user},
				fieldname: ["company"]
			},
			callback: function(r) {
				if(!r.exc && r.message && !in_list(["Administrator", "Guest"], frappe.session.user)){
					if(has_common(frappe.user_roles, ["Vlcc Operator", "Vlcc Manager"])){
						// $('body').find("[data-fieldname=vlcc]").val(r.message.company)
						frappe.query_report_filters_by_name.vlcc.set_input(r.message.company);
					}
					query_report.trigger_refresh();
				}
			}
		})
	},
	"formatter":function (row, cell, value, columnDef, dataContext, default_formatter) {
		value = default_formatter(row, cell, value, columnDef, dataContext);
        //console.log("concad", dataContext)
	    if(columnDef.id == "Explanation") {
	    	//console.log(dataContext,"customer customer")
            value = "<input type='text' name='fname'>";
	    }
	    return value;
	}
}

File: dairy_erp/dairy_erp/report/cattle_feed_sales_report/cattle_feed_sales_report.py
# Copyright (c) 2013, indictrans technologies and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import has_common

def execute(filters=None):
	columns, data = get_columns(), get_data(filters)
	return columns ,data 

def get_columns():

	columns = [
		_("Date") + ":Data:90", 
		_("Sales Invoice") + ":Link/Sales Invoice:150",
		_("Item Code") + ":Data:150",
		_("Quantity") + ":Float:150",
		_("Rate") + ":Currency:150",
		_("Total") + ":Currency:100",
		_("Remarks") + ":Data:200"
	]

	return columns

def get_data(filters):
	data = frappe.db.sql("""select
									DATE_FORMAT(si.posting_date, "%d-%m-%y"), 
									si.name,
									si_item.item_name,
									si_item.qty,
									si_item.rate,
									si_item.qty*si_item.rate 
							from
								`tabSales Invoice` si,
								`tabSales Invoice Item` si_item
							where
								si.local_sale = 1
								and si.name = si_item.parent
								and si.customer_or_farmer = "Farmer"
								and si_item.item_code not in ('COW Milk','BUFFALO Milk')
								and si.docstatus = 1 and si.company = '{0}'
								{1} order by si.posting_date """.format(filters.get('vlcc'),get_conditions(filters)),as_list=1,debug=0)
	if data:
		g_total = 0
		for row in data:
			g_total += row[5]
		data.append(["","","Grand Total","","",g_total,""])	
	return data

def get_conditions(filters):
	conditions = " and 1=1"
	if filters.get('farmer') and filters.get('start_date') and filters.get('end_date'):	
		conditions += " and si.posting_date between '{0}' and '{1}' and si.farmer = '{2}'".format(filters.get('start_date'),filters.get('end_date'),filters.get('farmer'))
	elif filters.get('start_date') and filters.get('end_date'):
		conditions += " and si.posting_date between '{0}' and '{1}' ".format(filters.get('start_date'),filters.get('end_date'))
	return conditions


